File: tools/anatomy_diagnostics.cpp
#include "simulation.hpp"

#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <string>

namespace {

void writeTriangle(std::ostream& out, const rp::World& world, const rp::Triangle& triangle, const char* fill, double opacity) {
    const auto& points = world.points();
    const rp::Vec2 a = points[triangle.a].position;
    const rp::Vec2 b = points[triangle.b].position;
    const rp::Vec2 c = points[triangle.c].position;
    out << "<polygon points=\""
        << a.x << "," << a.y << " "
        << b.x << "," << b.y << " "
        << c.x << "," << c.y
        << "\" fill=\"" << fill << "\" fill-opacity=\"" << opacity
        << "\" stroke=\"" << fill << "\" stroke-opacity=\"0.28\" stroke-width=\"1\"/>\n";
}

rp::Vec2 bonePoint(const rp::BoneSegment& bone, double t) {
    return {
        bone.a.x + (bone.b.x - bone.a.x) * t,
        bone.a.y + (bone.b.y - bone.a.y) * t,
    };
}

bool writeSvg(const std::filesystem::path& path, const rp::World& world, double width, double height) {
    if (path.has_parent_path()) {
        std::filesystem::create_directories(path.parent_path());
    }

    std::ofstream out(path);
    if (!out) {
        return false;
    }

    out << std::fixed << std::setprecision(2);
    out << "<svg xmlns=\"http://www.w3.org/2000/svg\" width=\"" << width << "\" height=\"" << height
        << "\" viewBox=\"0 0 " << width << " " << height << "\">\n";
    out << "<rect width=\"100%\" height=\"100%\" fill=\"#181818\"/>\n";

    for (const rp::Triangle& triangle : world.triangles()) {
        if (triangle.layer == rp::TissueLayer::Skin && world.triangleAlive(triangle)) {
            writeTriangle(out, world, triangle, "#9b705e", 0.22);
        }
    }
    for (const rp::Triangle& triangle : world.triangles()) {
        if (triangle.layer == rp::TissueLayer::Muscle && world.triangleAlive(triangle)) {
            writeTriangle(out, world, triangle, "#b62e3a", 0.50);
        }
    }

    for (const rp::BoneAttachment& attachment : world.boneAttachments()) {
        if (attachment.broken || attachment.bone >= world.bones().size() || attachment.point >= world.points().size()) {
            continue;
        }
        const rp::BoneSegment& bone = world.bones()[attachment.bone];
        const rp::Vec2 anchor = bonePoint(bone, attachment.t);
        const rp::Vec2 point = world.points()[attachment.point].position;
        out << "<line x1=\"" << anchor.x << "\" y1=\"" << anchor.y
            << "\" x2=\"" << point.x << "\" y2=\"" << point.y
            << "\" stroke=\"#57a6ff\" stroke-opacity=\"0.16\" stroke-width=\"1\"/>\n";
    }

    for (const rp::BoneJoint& joint : world.boneJoints()) {
        if (joint.broken || joint.a >= world.bones().size() || joint.b >= world.bones().size()) {
            continue;
        }
        const rp::Vec2 a = bonePoint(world.bones()[joint.a], joint.tA);
        const rp::Vec2 b = bonePoint(world.bones()[joint.b], joint.tB);
        out << "<line x1=\"" << a.x << "\" y1=\"" << a.y
            << "\" x2=\"" << b.x << "\" y2=\"" << b.y
            << "\" stroke=\"#ffd36a\" stroke-opacity=\"0.42\" stroke-width=\"2\"/>\n";
        out << "<circle cx=\"" << a.x << "\" cy=\"" << a.y << "\" r=\"3\" fill=\"#ffd36a\" fill-opacity=\"0.82\"/>\n";
        out << "<circle cx=\"" << b.x << "\" cy=\"" << b.y << "\" r=\"3\" fill=\"#ffd36a\" fill-opacity=\"0.82\"/>\n";
    }

    for (const rp::BoneSegment& bone : world.bones()) {
        out << "<line x1=\"" << bone.a.x << "\" y1=\"" << bone.a.y
            << "\" x2=\"" << bone.b.x << "\" y2=\"" << bone.b.y
            << "\" stroke=\"" << (bone.fractured ? "#fff3d6" : "#e5d5aa")
            << "\" stroke-width=\"" << bone.radius * 1.8
            << "\" stroke-linecap=\"round\"/>\n";
    }

    for (const rp::BoneSegment& bone : world.bones()) {
        constexpr int samples = 16;
        for (int i = 0; i < samples; ++i) {
            const double t = static_cast<double>(i) / static_cast<double>(samples - 1);
            const rp::Vec2 sample = bonePoint(bone, t);
            const bool insideSkin = rp::pointInsideLayer(world, sample, rp::TissueLayer::Skin);
            out << "<circle cx=\"" << sample.x << "\" cy=\"" << sample.y << "\" r=\"2.5\" fill=\""
                << (insideSkin ? "#59d36d" : "#ff4f42") << "\"/>\n";
        }
    }

    out << "</svg>\n";
    return true;
}

} // namespace

int main(int argc, char** argv) {
    constexpr double width = 1280.0;
    constexpr double height = 720.0;
    const std::filesystem::path output = argc > 1 ? std::filesystem::path(argv[1]) : std::filesystem::path("output/anatomy_debug.svg");

    const rp::World world = rp::createLayeredBody(width, height);
    const rp::AnatomyValidation validation = rp::validateAnatomy(world);

    std::cout << "points=" << world.points().size()
              << " springs=" << world.springs().size()
              << " triangles=" << world.triangles().size()
              << " bones=" << world.bones().size()
              << " bone_joints=" << world.boneJoints().size()
              << " bone_attachments=" << world.boneAttachments().size()
              << '\n';
    std::cout << "skin_points=" << validation.skinPoints
              << " muscle_points=" << validation.musclePoints
              << " bone_samples=" << validation.boneSamples
              << " bone_samples_outside_skin=" << validation.boneSamplesOutsideSkin
              << " bone_samples_outside_muscle=" << validation.boneSamplesOutsideMuscle
              << '\n';

    if (!writeSvg(output, world, width, height)) {
        std::cerr << "failed to write " << output.string() << '\n';
        return 1;
    }

    std::cout << "wrote " << output.string() << '\n';
    return validation.boneSamplesOutsideSkin == 0 ? 0 : 2;
}

File: src/bin/anatomy_diagnostics.rs
use realistic_physics as rp;
use std::env;
use std::fs;
use std::path::PathBuf;

fn main() {
    let output = env::args()
        .nth(1)
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from("output/anatomy_debug.svg"));
    let width = 1280.0;
    let height = 720.0;
    let world = rp::create_layered_body(width, height, rp::Materials::default());
    let validation = rp::validate_anatomy(&world, 16);

    let rib_count = world
        .bones()
        .iter()
        .filter(|bone| bone.kind == rp::BoneKind::Rib)
        .count();
    println!(
        "points={} springs={} triangles={} bones={} ribs={} bone_joints={} bone_attachments={} vessels={} cavities={} organs={}",
        world.points().len(),
        world.springs().len(),
        world.triangles().len(),
        world.bones().len(),
        rib_count,
        world.bone_joints().len(),
        world.bone_attachments().len(),
        world.vessels().len(),
        world.cavities().len(),
        world.organs().len()
    );
    println!(
        "skin_points={} muscle_points={} bone_samples={} bone_samples_outside_skin={} bone_samples_outside_muscle={}",
        validation.skin_points,
        validation.muscle_points,
        validation.bone_samples,
        validation.bone_samples_outside_skin,
        validation.bone_samples_outside_muscle
    );

    if let Some(parent) = output.parent() {
        if !parent.as_os_str().is_empty() {
            fs::create_dir_all(parent).expect("create output directory");
        }
    }
    fs::write(&output, build_svg(&world, width, height)).expect("write anatomy SVG");
    println!("wrote {}", output.display());

    if validation.bone_samples_outside_skin == 0 {
        std::process::exit(0);
    }
    std::process::exit(2);
}

fn build_svg(world: &rp::World, width: f64, height: f64) -> String {
    let mut out = String::new();
    out.push_str(&format!(
        "<svg xmlns=\"http://www.w3.org/2000/svg\" width=\"{width:.2}\" height=\"{height:.2}\" viewBox=\"0 0 {width:.2} {height:.2}\">\n"
    ));
    out.push_str("<rect width=\"100%\" height=\"100%\" fill=\"#181818\"/>\n");

    for triangle in world.triangles() {
        if triangle.layer == rp::TissueLayer::Skin && world.triangle_alive(triangle) {
            write_triangle(&mut out, world, triangle, "#9b705e", 0.22);
        }
    }
    for triangle in world.triangles() {
        if triangle.layer == rp::TissueLayer::Muscle && world.triangle_alive(triangle) {
            write_triangle(&mut out, world, triangle, "#b62e3a", 0.50);
        }
    }
    for vessel in world.vessels() {
        out.push_str(&format!(
            "<line x1=\"{:.2}\" y1=\"{:.2}\" x2=\"{:.2}\" y2=\"{:.2}\" stroke=\"#b51224\" stroke-opacity=\"0.72\" stroke-width=\"{:.2}\" stroke-linecap=\"round\"/>\n",
            vessel.a.x,
            vessel.a.y,
            vessel.b.x,
            vessel.b.y,
            vessel.radius * 1.35 + 1.0
        ));
    }
    for attachment in world.bone_attachments() {
        if attachment.broken
            || attachment.bone >= world.bones().len()
            || attachment.point >= world.points().len()
        {
            continue;
        }
        let anchor = bone_point(world.bones()[attachment.bone], attachment.t);
        let point = world.points()[attachment.point].position;
        out.push_str(&format!(
            "<line x1=\"{:.2}\" y1=\"{:.2}\" x2=\"{:.2}\" y2=\"{:.2}\" stroke=\"#57a6ff\" stroke-opacity=\"0.16\" stroke-width=\"1\"/>\n",
            anchor.x, anchor.y, point.x, point.y
        ));
    }
    for joint in world.bone_joints() {
        if joint.broken || joint.a >= world.bones().len() || joint.b >= world.bones().len() {
            continue;
        }
        let a = bone_point(world.bones()[joint.a], joint.t_a);
        let b = bone_point(world.bones()[joint.b], joint.t_b);
        out.push_str(&format!(
            "<line x1=\"{:.2}\" y1=\"{:.2}\" x2=\"{:.2}\" y2=\"{:.2}\" stroke=\"#ffd36a\" stroke-opacity=\"0.42\" stroke-width=\"2\"/>\n",
            a.x, a.y, b.x, b.y
        ));
        out.push_str(&format!(
            "<circle cx=\"{:.2}\" cy=\"{:.2}\" r=\"3\" fill=\"#ffd36a\" fill-opacity=\"0.82\"/>\n",
            a.x, a.y
        ));
        out.push_str(&format!(
            "<circle cx=\"{:.2}\" cy=\"{:.2}\" r=\"3\" fill=\"#ffd36a\" fill-opacity=\"0.82\"/>\n",
            b.x, b.y
        ));
    }
    for bone in world.bones() {
        let intact_stroke = if bone.kind == rp::BoneKind::Rib {
            "#f0dfb6"
        } else {
            "#e5d5aa"
        };
        out.push_str(&format!(
            "<line x1=\"{:.2}\" y1=\"{:.2}\" x2=\"{:.2}\" y2=\"{:.2}\" stroke=\"{}\" stroke-width=\"{:.2}\" stroke-linecap=\"round\"/>\n",
            bone.a.x,
            bone.a.y,
            bone.b.x,
            bone.b.y,
            if bone.fractured { "#fff3d6" } else { intact_stroke },
            bone.radius * 1.8
        ));
    }
    for bone in world.bones() {
        for i in 0..16 {
            let t = i as f64 / 15.0;
            let sample = bone_point(*bone, t);
            let inside_skin = rp::point_inside_layer(world, sample, rp::TissueLayer::Skin);
            out.push_str(&format!(
                "<circle cx=\"{:.2}\" cy=\"{:.2}\" r=\"2.5\" fill=\"{}\"/>\n",
                sample.x,
                sample.y,
                if inside_skin { "#59d36d" } else { "#ff4f42" }
            ));
        }
    }
    out.push_str("</svg>\n");
    out
}

fn write_triangle(
    out: &mut String,
    world: &rp::World,
    triangle: &rp::Triangle,
    fill: &str,
    opacity: f64,
) {
    let points = world.points();
    let a = points[triangle.a].position;
    let b = points[triangle.b].position;
    let c = points[triangle.c].position;
    out.push_str(&format!(
        "<polygon points=\"{:.2},{:.2} {:.2},{:.2} {:.2},{:.2}\" fill=\"{}\" fill-opacity=\"{:.2}\" stroke=\"{}\" stroke-opacity=\"0.28\" stroke-width=\"1\"/>\n",
        a.x, a.y, b.x, b.y, c.x, c.y, fill, opacity, fill
    ));
}

fn bone_point(bone: rp::BoneSegment, t: f64) -> rp::Vec2 {
    rp::Vec2 {
        x: bone.a.x + (bone.b.x - bone.a.x) * t,
        y: bone.a.y + (bone.b.y - bone.a.y) * t,
    }
}
